src/porepy/compositional/materials.py

```diff
@@ -0,0 +1,536 @@
+"""Storage classes for material constants.
+
+Materials are storage classes for values of physical properties. They are typically used
+when composing constitutive laws. A material is instantiated with a Units object, which
+defines the units of the physical properties. The material can then be used to convert
+from units set by the user (standard SI units) to those specified by the Units object,
+which are the ones used internally in the simulation. The conversion hopefully reduces
+problems with scaling/rounding errors and condition numbers.
+
+"""
+
+from __future__ import annotations
+
+from collections import OrderedDict
+from dataclasses import asdict, dataclass, field, is_dataclass
+from typing import Callable, ClassVar, Generic, Optional, TypeVar, Union, cast, overload
+
+import numpy as np
+
+import porepy as pp
+
+__all__ = [
+    "MaterialConstants",
+    "FluidConstants",
+    "SolidConstants",
+    "load_fluid_constants",
+]
+
+number = pp.number
+
+_K = TypeVar("_K")
+_V = TypeVar("_V")
+
+
+class _HashableDict(Generic[_K, _V], OrderedDict):
+    """See https://stackoverflow.com/questions/1151658/python-hashable-dicts.
+
+    We require hashable dictionaries for the below material constant classes which
+    contain various constants and unit declarations in dicts, all in simple formats and per se
+    hashable.
+
+    The need for hashable material constants arises when fluid constants are used as a base
+    class for components and tracers, which are dynamically created and themselves stored in
+    standard dictionaries at various points in ``porepy.compositional``.
+
+    """
+
+    def __hash__(self):
+        return hash(tuple(sorted(self.items())))
+
+
+# The dataclass is frozen, since material parameters are not supposed to change, as well
+# as the units they are given in.
+# By using this in combination with keyword_only arguments for the construction of
+# materials, the user is forced to instantiate the constants with the right names of
+# constants, otherwise errors are raised (no separate checks required).
+# By providing default values to the fields, not every constant is required. The user
+# is expected to be aware which physics are used in the model.
+@dataclass(frozen=True, kw_only=True)
+class MaterialConstants:
```

Comment:
Yes. The original implementation performed a conversion when the constant was called, IF it had a physical dimension. Otherwise it just returned the constant.
Now the post init processing takes care of the conversion, IF a dimension is given in SI_units, and we can define constants simply using the dataclass mechanism.